### PR TITLE
Improve Sprite Misc Update

### PR DESCRIPTION
**Pull request by duncanspumpkin**
I've added member functions for all of the misc sprites to improve the code. I've also removed the reinterpret cast and now use the safe conversion functions for all of the types.

### CODE REVIEW

File: src/openrct2/world/Particle.cpp [hunk 1]
```diff
@@ -57,63 +57,61 @@ void crashed_vehicle_particle_create(rct_vehicle_colour colours, const CoordsXYZ
...
+    velocity_z = vz & 0xFFFF;
 
     // Check collision with land / water
-    int16_t landZ = tile_element_height({ x, y });
-    int16_t waterZ = tile_element_water_height({ x, y });
+    int16_t landZ = tile_element_height(newLoc);
+    int16_t waterZ = tile_element_water_height(newLoc);
 
-    if (waterZ != 0 && particle->z >= waterZ && z <= waterZ)
+    if (waterZ != 0 && z >= waterZ && z <= waterZ)
```
Comment (duncanspumpkin) [inline src/openrct2/world/Particle.cpp:93]: Second z should be newLoc

File: src/openrct2/world/Particle.cpp [hunk 2]
```diff
@@ -57,63 +57,61 @@ void crashed_vehicle_particle_create(rct_vehicle_colour colours, const CoordsXYZ
...
-        crash_splash_create({ particle->x, particle->y, waterZ });
-        sprite_remove(particle);
+        audio_play_sound_at_location(SoundId::Water2, { x, y, waterZ });
+        crash_splash_create({ x, y, waterZ });
+        sprite_remove(this);
         return;
     }
 
-    if (particle->z >= landZ && z <= landZ)
+    if (z >= landZ && z <= landZ)
```
Comment (duncanspumpkin) [inline src/openrct2/world/Particle.cpp:102]: Second z should be newLoc.z

File: src/openrct2/world/Sprite.cpp [hunk 3]
```diff
@@ -607,51 +607,51 @@ void sprite_misc_explosion_flare_create(const CoordsXYZ& flarePos)
...
     }
 }
 
 /**
  *
  *  rct2: 0x006731CD
  */
-static void sprite_misc_update(rct_sprite* sprite)
+static void sprite_misc_update(SpriteBase* sprite)
 {
```
Comment (tupaschoal) [inline src/openrct2/world/Sprite.cpp:625]: Should this function be checking for pointer validity?

File: src/openrct2/world/Particle.cpp (hunk 1 above)
Comment (tupaschoal) [reply src/openrct2/world/Particle.cpp:93]:
```suggestion
    if (waterZ != 0 && z >= waterZ && newLoc.z <= waterZ)
```

File: src/openrct2/world/Particle.cpp (hunk 2 above)
Comment (tupaschoal) [reply src/openrct2/world/Particle.cpp:102]:
```suggestion
    if (z >= landZ && newLoc.z <= landZ)
```

File: src/openrct2/world/Sprite.h [hunk 4]
```diff
@@ -61,15 +61,18 @@ struct Duck : SpriteGeneric
     int16_t target_y;
     uint8_t state;
 
+    void Update();
+    uint32_t GetFrameImage(int32_t direction) const;
+    void Invalidate();
+    bool IsFlying();
+    void Remove();
```
Comment (tupaschoal) [inline src/openrct2/world/Sprite.h:68]: Why unprotecting all of these in this changeset?

Comment (duncanspumpkin) [reply src/openrct2/world/Sprite.h:68]: They were always unprotected as it's a struct. I added protection to the update sub functions though as they no longer have to be public.

Comment (tupaschoal) [reply src/openrct2/world/Sprite.h:68]: Ah, I see, I had missed that `private` was one of the additions as well

File: src/openrct2/world/Sprite.cpp (hunk 3 above)
Comment (duncanspumpkin) [reply src/openrct2/world/Sprite.cpp:625]: I'm hoping that that is not required as the As template should mean that part of it is taken care of by the switch. None of the compilers are complaining about that so i think it is happening like that.

Comment (duncanspumpkin) [reply src/openrct2/world/Sprite.cpp:625]: Well i checked the assembly of the version generated on my PC. It did not work out that its already validated the type so not quite as good as I hoped.

File: src/openrct2/world/Sprite.h (hunk 4 above)
Comment (AaronVanGeffen) [reply src/openrct2/world/Sprite.h:68]: If you're adding private members, this should probably be a class instead.

Comment (duncanspumpkin) [reply src/openrct2/world/Sprite.h:68]: I was adding a public member. It's just the 4 existing public members no longer need to be public.

**Pull request merged**